lighthouse-core/computed/page-dependency-graph.js

```diff
@@ -151,8 +151,10 @@ class PageDependencyGraph {
       if (initiators.length) {
         initiators.forEach(initiator => {
           const parentCandidates = networkNodeOutput.urlToNodeMap.get(initiator) || [];
-          // Only add the edge if the parent is unambiguous with valid timing.
-          if (parentCandidates.length === 1 && parentCandidates[0].startTime <= node.startTime) {
+          // Only add the edge if the parent is unambiguous with valid timing and isn't circular.
+          if (parentCandidates.length === 1 &&
+              parentCandidates[0].startTime <= node.startTime &&
+              !parentCandidates[0].isDependentOn(node)) {
```

Comment:
should we be doing the same/similar fixes in `networkRecorder._chooseInitiator`? I can see an argument for "everyone should be using `PageDependencyGraph`", but if we're going to have `networkRequest.initiator` exposed, we should want it to be sensical?

Reply:
> `networkRecorder._chooseInitiator`

actually, the relationship of `_chooseInitiator` with `PageDependencyGraph.getNetworkInitiators` is a little confusing. If seems like `networkRecorder._chooseInitiator` tries to get an initiator however it can, including from the top of `stack`, but that seems like it would short circuit `PageDependencyGraph.getNetworkInitiators` with an early `[record.initiator.url]` return and it would never look into `stack`, even if `initiator.type` is `script`?

Is that desired or am I missing how that works?

Reply:
> should we be doing the same/similar fixes in networkRecorder._chooseInitiator?

I don't think so. We really don't have a way of piecing together the full graph picture in networkRecorder without duplicating page-depedency-graph and we have to have *some* layer that represents what Chrome said. I'm OK keeping the request the (mostly) raw info from the protocol and filtering  out information that's conflicting *in aggregate* at this graph layer.

Reply:
> If seems like networkRecorder._chooseInitiator tries to get an initiator however it can, including from the top of stack, but that seems like it would short circuit PageDependencyGraph.getNetworkInitiators with an early [record.initiator.url] return and it would never look into stack, even if initiator.type is script?

That won't happen because in `page-dependency-graph` we look at `.initiators` *first* and fallback to `.initiatorRequest` only if we can't find the requests from initiators. That behavior wasn't tested though so added a test for it 👍 

This also brings up a good point that in network-recorder we don't have the same caution with ambiguity though, so added a test for that as well 👍

Reply:
ugh, `initiator` vs `initiatorRequest`, right...